compiler/rustc_mir/src/transform/copy_propagation.rs

```diff
@@ -0,0 +1,298 @@
+        }
+
+        // Ensure that src has an entry so that we can detect invalidation.
+        let src_generation =
+            self.values.entry(src.local).or_insert(LocalValue::default()).generation;
+
+        // Record `dst = src` assignment.
+        let value = &mut self.values.entry(dst).or_insert(LocalValue::default());
+        value.generation.0 += 1;
+        value.src = Some(*src);
+        value.src_generation = src_generation;
+    }
+
+    /// Replaces a use of dst with its current value.
+    fn propagate_local(&mut self, dst: &mut Local) {
+        let dst_value = match self.values.get(dst) {
+            None => return,
+            Some(dst_value) => dst_value,
+        };
+        let src = match dst_value.src {
+            Some(src) => src,
+            None => return,
+        };
+        let src = match src.as_local() {
+            Some(src) => src,
+            None => return,
+        };
+        // Last definition of dst was of the form `dst = src`.
+
+        // Check that src still has the same value.
+        if dst_value.src_generation != self.values.get(&src).unwrap().generation {
+            return;
+        }
+
+        // Propagate
+        *dst = src;
+    }
+
+    /// Replaces a use of dst with its current value.
+    fn propagate_place(&mut self, dst: &mut Place<'tcx>) {
+        let dst_value = match self.values.get(&dst.local) {
+            None => return,
+            Some(dst_value) => dst_value,
+        };
+
+        let src = match dst_value.src {
+            Some(src) => src,
+            None => return,
+        };
+        // Last definition of dst.local was of the form `dst.local = src`.
+
+        // Check that src still has the same value.
+        if dst_value.src_generation != self.values.get(&src.local).unwrap().generation {
+            return;
+        }
+
+        // Propagate while rebasing any additional projections on top of src.
+        *dst = Place {
+            local: src.local,
+            projection: rebase_onto(self.tcx, dst.projection, src.projection),
+        };
+    }
+}
+
+/// Rebase projection `proj` on top of a `base`.
+fn rebase_onto<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    proj: &'tcx List<PlaceElem<'tcx>>,
+    base: &'tcx List<PlaceElem<'tcx>>,
+) -> &'tcx List<PlaceElem<'tcx>> {
+    if base.is_empty() {
+        return proj;
+    }
+    if proj.is_empty() {
+        return base;
+    }
+    let mut projection: Vec<_> = Vec::with_capacity(base.len() + proj.len());
+    projection.extend(base);
+    projection.extend(proj);
+    tcx.intern_place_elems(&projection)
+}
+
+/// Invalidates locals that could be modified during execution of visited MIR.
+struct InvalidateModifiedLocals<'tcx, 'a> {
+    values: &'a mut LocalValues<'tcx>,
+}
+
+impl<'tcx, 'a> Visitor<'tcx> for InvalidateModifiedLocals<'tcx, 'a> {
+    fn visit_local(&mut self, local: &Local, context: PlaceContext, _location: Location) {
+        match context {
+            PlaceContext::NonUse(_)
```

Comment:
This shouldn't invalidate the local, right?

Reply:
Non-use context includes storage live and storage dead (and others that shouldn't be encountered there). It would make sense to expand this for clarity.